man/faq/internal/ptype2-identity.Rmd

```diff
@@ -0,0 +1,81 @@
+
+```{r, child = "setup.Rmd", include = FALSE}
+```
+
+## Promotion monoid
+
+`vec_ptype2()` returns the _promotion_ type of two vectors, if any. This is typically the richer type of the two so that automatic coercions are never lossy. For example, the promotion type of integer and double is the latter because double covers a larger range of values than integer.
+
+`vec_ptype2()` is a [monoid](https://en.wikipedia.org/wiki/Monoid) over vectors, which in practical terms means that it is a well behaved operation for [reduction](https://purrr.tidyverse.org/reference/reduce.html). As such it needs an identity element, i.e. a value that doesn't change the result of the reduction. vctrs has two such identity values, `NULL` and __unspecified__ vectors.
+
+
+## The `NULL` identity
+
+As an identity element that shouldn't influence the determination of the common type of a set of vectors, `NULL` is promoted to any type:
+
+```{r}
+vec_ptype2(NULL, "")
+vec_ptype2(1L, NULL)
+```
+
+The common type of `NULL` and `NULL` is the identity `NULL`:
+
+```{r}
+vec_ptype2(NULL, NULL)
+```
+
+This way the result of `vec_ptype2(NULL, NULL)` does not influence subsequent promotions:
+
+```{r}
+vec_ptype2(
+  vec_ptype2(NULL, NULL),
+  ""
+)
+```
+
+
+## Unspecified vectors
+
+In the vctrs coercion system, logical vectors of missing values are also automatically promoted to the type of any other vector, just like `NULL`. We call these vectors unspecified. The special coercion semantics of unspecified vectors serve two purposes:
+
+1. It makes it possible to assign vectors of `NA` inside any type of vectors, even when they are not coercible with logical:
```

Comment:
Is this even for foreign classes? We had a test where `NA` assignment fails for `foobar()`, and I added another one that is highly specific to unspecified types here https://github.com/r-lib/vctrs/pull/812/files#diff-51c944df297abe15044974191b4f4ebbR271-R308

Practically, the problem with allowing any foreign type is that we have to cast `NA` to `<foreign-type>` at some point. I think ideally we'd just call `vec_init(to, vec_size(x))`, but this doesn't work for all types, i.e. integer64, which doesn't implement `NA_integer_` slicing in the CRAN version.

Reply:
This should work for foreign classes as well if they implement ptype2 correctly. `foobar` is not a real class so the methods might be buggy in tests, and need to be adjusted sometimes. As for integer64, it is buggy. `vec_init()` is a separate issue though, this is about ptype2 and unspecified, and perhaps proxy/restore.

Reply:
oh now I see what you mean about init, it's used in `vec_default_cast()` for the default coercion of unspecified to the `to` type.